File: ios/rgb/RGBModule.m
//
//  RGBModule.m
//  HEXA
//
//  Created by Shashank Shinde on 26/06/23.
//


#import "RGBModule.h"
#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>
#import "HEXA-Swift.h"

#import <Foundation/Foundation.h>


@implementation RGB

RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(generateKeys:(NSString*)network
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  
  [helper generateKeysWithBtcNetwotk:network callback:^(NSString * _Nonnull response) {
    resolve(response);
  }];
}

RCT_EXPORT_METHOD(getAddress:(NSString*)mnemonic
                  network:(NSString *)network
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  [helper getAddressWithBtcNetwotk:network mnemonic:mnemonic callback:^(NSString * _Nonnull response) {
      resolve(response);
    }
   ];
}

RCT_EXPORT_METHOD(initiate:(NSString*)network
                  mnemonic:(NSString *)mnemonic
                  pubKey:(NSString *)pubKey
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  [helper initiateWithBtcNetwotk:network mnemonic:mnemonic pubkey:pubKey callback:^(NSString * _Nonnull response) {
    resolve(response);
  }
   ];
}

RCT_EXPORT_METHOD(sync:(NSString*)mnemonic
                  network:(NSString *)network
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  [helper syncWithBtcNetwotk:network mnemonic:mnemonic callback:^(NSString * _Nonnull response) {
    resolve(response);
  }
   ];
}

RCT_EXPORT_METHOD(getBalance:(NSString*)mnemonic
                  network:(NSString *)network
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  [helper getBalanceWithBtcNetwotk:network mnemonic:mnemonic callback:^(NSString * _Nonnull response) {
      resolve(response);
    }
   ];
}

RCT_EXPORT_METHOD(getTransactions:(NSString*)mnemonic
                  network:(NSString *)network
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  [helper getTransactionsWithBtcNetwotk:network mnemonic:mnemonic callback:^(NSString * _Nonnull response) {
      resolve(response);
    }
   ];
}

RCT_EXPORT_METHOD(syncRgbAssets:(NSString*)mnemonic
                  pubKey:(NSString *)pubKey
                  network:(NSString *)network
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  [
    helper syncRgbWithCallback:^(NSString * _Nonnull response) {
      resolve(response);
    }
   ];
}

RCT_EXPORT_METHOD(receiveAsset:(NSString*)mnemonic
                  network:(NSString *)network
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  [
    helper receiveAssetWithBtcNetwotk:network mnemonic:mnemonic callback:^(NSString * _Nonnull response) {
      resolve(response);
    }
   ];
}


RCT_EXPORT_METHOD(getRgbAssetMetaData:(NSString*)assetId
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  [
    helper getRgbAssetMetaDataWithAssetId:assetId callback:^(NSString * _Nonnull response) {
      resolve(response);
    }
   ];
}


RCT_EXPORT_METHOD(getRgbAssetTransactions:(NSString*)assetId
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  [helper getRgbAssetTransactionsWithAssetId:assetId callback:^(NSString * _Nonnull response) {
    resolve(response);
  }
   ];
}

RCT_EXPORT_METHOD(sendBtc:(NSString*)mnemonic
                  network:(NSString *)network
                  address:(NSString *)address
                  amount:(NSString *)amount
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  [
    helper sendBtcWithBtcNetwotk:network mnemonic:mnemonic address:address amount:amount callback:^(NSString * _Nonnull response) {
      resolve(response);
    }
   ];
}

RCT_EXPORT_METHOD(issueRgb20Asset:(NSString*)ticker
                  name:(NSString *)name
                  supply:(NSString *)supply
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];

  [helper issueRgb20AssetWithTicker:ticker name:name supply:supply callback:^(NSString * _Nonnull response) {
    resolve(response);
  }];
}

RCT_EXPORT_METHOD(issueRgb25Asset:(NSString*)description
                  name:(NSString *)name
                  supply:(NSString *)supply
                  filePath:(NSString *)filePath
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  
  [helper issueRgb25AssetWithName:name description:description supply:supply filePath:filePath callback:^(NSString * _Nonnull response) {
    resolve(response);
  }
   ];
}

RCT_EXPORT_METHOD(sendAsset:(NSString*)assetId
                  blindedUTXO:(NSString *)blindedUTXO
                  amount:(NSString *)amount
                  consignmentEndpoints:(NSString *)consignmentEndpoints
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject){
  RGBHelper *helper = [[RGBHelper alloc]init];
  
  [helper sendAssetWithAssetId:assetId blindedUtxo:blindedUTXO amount:amount consignmentEndpoints:consignmentEndpoints callback:^(NSString * _Nonnull response) {
    resolve(response);
  }
   ];
}

@end

File: android/app/src/main/java/io/hexawallet/hexa/RGBModule.kt
package io.hexawallet.hexa
import android.util.Log
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Promise
import com.google.gson.Gson
import com.google.gson.JsonObject
import org.rgbtools.BitcoinNetwork
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper

import kotlin.math.log

class RGBModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    val TAG = "RGBMODULE"
    private val backgroundHandler: Handler

    init {
        val handlerThread = HandlerThread("BackgroundThread")
        handlerThread.start()
        backgroundHandler = Handler(handlerThread.looper)
    }
    override
    fun getName() = "RGB"

    private fun getRgbNetwork(network: String): BitcoinNetwork {
        return if (network == "TESTNET") BitcoinNetwork.TESTNET else BitcoinNetwork.MAINNET
    }

    @ReactMethod
    fun generateKeys(network: String, promise: Promise) {
        val rgbNetwork = if(network == "TESTNET") BitcoinNetwork.TESTNET else BitcoinNetwork.MAINNET
        val keys = org.rgbtools.generateKeys(rgbNetwork)
        val gson = Gson()
        val json = gson.toJson(keys)
        promise.resolve(json.toString())
    }

    @ReactMethod
    fun restoreKeys(network: String, mnemonic: String, promise: Promise) {
        val rgbNetwork = if(network == "TESTNET") BitcoinNetwork.TESTNET else BitcoinNetwork.MAINNET
        val keys = org.rgbtools.restoreKeys(rgbNetwork, mnemonic)
        val gson = Gson()
        val json = gson.toJson(keys)
        promise.resolve(json.toString())
    }

    @ReactMethod
    fun getAddress(promise: Promise) {
        promise.resolve(RGBHelper.getAddress())
    }
    @ReactMethod
    fun initiate(network: String, mnemonic:String, xpub: String, promise: Promise){
        promise.resolve(RGBWalletRepository.initialize(network,mnemonic,xpub))
    }

    @ReactMethod
    fun sync( mnemonic:String, network: String, promise: Promise){
    }

    @ReactMethod
    fun getBalance( mnemonic:String,network: String, promise: Promise){
        promise.resolve("")
    }

    @ReactMethod
    fun getTransactions( mnemonic:String,network: String, promise: Promise){
        promise.resolve("")
    }


    @ReactMethod
    fun syncRgbAssets( mnemonic:String, pubKey:String, network: String, promise: Promise){
        backgroundHandler.post{
            promise.resolve(RGBHelper.syncRgbAssets())
        }
    }

    @ReactMethod
    fun receiveAsset(promise: Promise){
        backgroundHandler.post{
            try {
                promise.resolve(RGBHelper.receiveAsset())
            }catch (e: Exception) {
                val message = e.message
                val jsonObject = JsonObject()
                jsonObject.addProperty("error", message)
                promise.resolve(jsonObject.toString())
            }
        }
    }

    @ReactMethod
    fun createUtxos(feeRate: Float, promise: Promise){
        backgroundHandler.post {
            try {
                Log.d(TAG, "createUtxos: $feeRate")
                val created = RGBHelper.createNewUTXOs(feeRate)
                val jsonObject = JsonObject()
                jsonObject.addProperty("error", "")
                jsonObject.addProperty("created", created)
                promise.resolve(jsonObject.toString())
            }catch (e: Exception) {
                val message = e.message
                val jsonObject = JsonObject()
                jsonObject.addProperty("error", message)
                promise.resolve(jsonObject.toString())
            }
        }

    }

    @ReactMethod
    fun issueRgb20Asset( ticker: String, name: String, supply: String, promise: Promise){
        backgroundHandler.post {
            Log.d(TAG, "issueRgb20Asset: ${supply}")
            try {
                val amounts = listOf(supply)
                val response = RGBHelper.issueRgb20Asset(ticker, name, amounts.map { it.toULong() })
                promise.resolve(response)
            }catch (e: Exception) {
                Log.d(TAG, "issueRgb20Asset:e.message ${e.message}")

                val message = e.message
                val jsonObject = JsonObject()
                jsonObject.addProperty("error", message)
                promise.resolve(jsonObject.toString())
            }
        }
    }

    @ReactMethod
    fun issueRgb25Asset( description: String, name: String, supply: String,filePath: String, promise: Promise){
        backgroundHandler.post{
            try {
                val amounts = listOf(supply)
                val response = RGBHelper.issueRgb25Asset(description, name, amounts.map { it.toULong() }, filePath)
                promise.resolve(response)
            }catch (e: Exception) {
                Log.d(TAG, "issueRgb20Asset:e.message ${e.message}")
                val message = e.message
                val jsonObject = JsonObject()
                jsonObject.addProperty("error", message)
                promise.resolve(jsonObject.toString())
            }
        }
    }

    @ReactMethod
    fun getRgbAssetMetaData( assetId: String, promise: Promise){
        promise.resolve(RGBHelper.getMetadata(assetId))
    }

    @ReactMethod
    fun getRgbAssetTransactions( assetId: String, promise: Promise){
        promise.resolve(RGBHelper.getAssetTransfers(assetId))
    }

    @ReactMethod
    fun sendAsset( assetId: String, blindedUTXO: String, amount: String, consignmentEndpoints: String, promise: Promise){
        backgroundHandler.post {
            try {
                val endpoints = listOf(consignmentEndpoints)
                Log.d(TAG, "sendAsset: blindedUTXO=$blindedUTXO")
                promise.resolve(RGBHelper.send(assetId, blindedUTXO, amount.toULong(), endpoints))
            }catch (e: Exception) {
                val message = e.message
                val jsonObject = JsonObject()
                jsonObject.addProperty("error", message)
                promise.resolve(jsonObject.toString())
            }
        }
    }

    @ReactMethod
    fun getUnspents(promise: Promise){
        val rgbUtxo = RGBHelper.getUnspents()
        //val bitcoinUtxo = BdkHelper.getUnspents()
        val jsonObject = JsonObject()
        jsonObject.addProperty("rgb", Gson().toJson(rgbUtxo))
        //jsonObject.addProperty("bitcoin", Gson().toJson(bitcoinUtxo))
        promise.resolve(jsonObject.toString())
    }

    @ReactMethod
    fun refreshAsset(assetId: String, promise: Promise){
        backgroundHandler.post {
            val rgbUtxo = RGBHelper.getUnspents()
            //val bitcoinUtxo = BdkHelper.getUnspents()
            val jsonObject = JsonObject()
            jsonObject.addProperty("rgb", Gson().toJson(rgbUtxo))
            //jsonObject.addProperty("bitcoin", Gson().toJson(bitcoinUtxo))
            promise.resolve(jsonObject.toString())
        }
    }
    @ReactMethod
    fun isValidBlindedUtxo(invoice:String,promise: Promise){
        val response = RGBHelper.isValidInvoice(invoice)
        promise.resolve(response)
    }

    @ReactMethod
    fun backup(backupPath: String, password: String, promise: Promise){
        backgroundHandler.post {
            val response = RGBHelper.backup(backupPath, password, reactApplicationContext)
            promise.resolve(response.toString())
        }
    }

    @ReactMethod
    fun isBackupRequired(promise: Promise){
        val response = RGBHelper.getBackupInfo()
        promise.resolve(response)
    }

    @ReactMethod
    fun restore(mnemonic: String,promise: Promise){
        val response = RGBHelper.restore(mnemonic, reactApplicationContext)
        Log.d(TAG, "restore: $response")
        promise.resolve(response)
    }

    @ReactMethod
    fun resetData(promise: Promise){
        AppConstants.rgbDir.delete()
        promise.resolve(true)
    }
}
